File: test/integration/python/FrugalTestHandler.py
import time

from frugal_test.f_FrugalTest import Iface, Xtruct, Xception, Xception2, Insanity


class FrugalTestHandler(Iface):
    def testVoid(self, ctx):
        print "test_void({})".format(ctx)
        return

    def testString(self, ctx, thing):
        print "test_string({})".format(thing)
        return thing

    def testBool(self, ctx, thing):
        print "test_bool({})".format(thing)
        return thing

    def testByte(self, ctx, thing):
        print "test_byte({})".format(thing)
        return thing

    def testI32(self, ctx, thing):
        print "test_i32({})".format(thing)
        return thing

    def testI64(self, ctx, thing):
        print "test_i64({})".format(thing)
        return thing

    def testDouble(self, ctx, thing):
        print "test_double({})".format(thing)
        return thing

    def testBinary(self, ctx, thing):
        print "test_binary({})".format(thing)
        return thing

    def testStruct(self, ctx, thing):
        print "test_struct({})".format(thing)
        return thing

    def testNest(self, ctx, thing):
        print "test_nest({})".format(thing)
        return thing

    def testMap(self, ctx, thing):
        print "test_map({})".format(thing)
        return thing

    def testStringMap(self, ctx, thing):
        print "test_string_map({})".format(thing)
        return thing

    def testSet(self, ctx, thing):
        print "test_set({})".format(thing)
        return thing

    def testList(self, ctx, thing):
        print "test_list({})".format(thing)
        return thing

    def testEnum(self, ctx, thing):
        print "test_enum({})".format(thing)
        return thing

    def testMapMap(self, ctx, hello):
        print "test_map_map({})".format(hello)
        d = {4: 4, 3: 3, 2: 2, 1: 1}
        e = {-4: -4, -3: -3, -2: -2, -1: -1}
        mapmap = {-4: e, 4: d}
        return mapmap

    def testInsanity(self, ctx, argument):
        print "test_insanity({})".format(argument)
        return {1:
                {2: argument,
                 3: argument}, 2: {}}

    def testMulti(self, ctx, arg0, arg1, arg2, arg3, arg4, arg5):
        print "test_multi({}, {}, {}, {}, {}, {})".format(arg0, arg1, arg2, arg3, arg4, arg5)
        result = Xtruct()
        result.string_thing = "Hello2"
        result.byte_thing = arg0
        result.i32_thing = arg1
        result.i64_thing = arg2
        return result

    def testException(self, ctx, arg):
        print "test_exception({})".format(arg)
        if arg == "Xception":
            e = Xception()
            e.errorCode = 1001
            e.message = arg
            raise e
        elif arg == "TException":
            raise Xception.message("Just TException")
        return

    def testMultiException(self, ctx, arg0, arg1):
        print "test_multi_exception({}, {})".format(arg0, arg1)
        if arg0 == "Xception":
            e = Xception()
            e.errorCode = 1001
            e.message = "This is an Xception"
            raise e
        elif arg0 == "Xception2":
            e = Xception2()
            e.errorCode = 2002
            e.struct_thing = Xtruct()
            e.struct_thing.string_thing = "This is an Xception2"
            raise e
        else:
            r = Xtruct()
            r.string_thing = arg1
            return r

    def testOneway(self, ctx, seconds):
        print "test_oneway({}): Sleeping...".format(seconds)
        time.sleep(seconds)
        print "testOneway({}): done sleeping!".format(seconds)
        return

File: test/integration/java/frugal-integration-test/src/main/java/com/workiva/FrugalTestHandler.java
package com.workiva;

import com.workiva.frugal.FContext;
import frugal.test.*;
import org.apache.thrift.TException;
import frugal.test.Numberz;

import java.nio.ByteBuffer;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;


public class FrugalTestHandler implements FFrugalTest.Iface {

        // Each RPC handler "test___" accepts a value of type ___ and returns the same value (where applicable).
        // The client then asserts that the returned value is equal to the value sent.
        @Override
        public void testVoid(FContext ctx) throws TException {
        }

        @Override
        public String testString(FContext ctx, String thing) throws TException {
            return thing;
        }

        @Override
        public boolean testBool(FContext ctx, boolean thing) throws TException {
            return thing;
        }

        @Override
        public byte testByte(FContext ctx, byte thing) throws TException {
            return thing;
        }

        @Override
        public int testI32(FContext ctx, int thing) throws TException {
            return thing;
        }

        @Override
        public long testI64(FContext ctx, long thing) throws TException {
            return thing;
        }

        @Override
        public double testDouble(FContext ctx, double thing) throws TException {
            return thing;
        }

        @Override
        public ByteBuffer testBinary(FContext ctx, ByteBuffer thing) throws TException {
            return thing;
        }

        @Override
        public Xtruct testStruct(FContext ctx, Xtruct thing) throws TException {
            return thing;
        }

        @Override
        public Xtruct2 testNest(FContext ctx, Xtruct2 thing) throws TException {
            return thing;
        }

        @Override
        public Map<Integer, Integer> testMap(FContext ctx, Map<Integer, Integer> thing) throws TException {
            return thing;
        }

        @Override
        public Map<String, String> testStringMap(FContext ctx, Map<String, String> thing) throws TException {
            return thing;
        }

        @Override
        public Set<Integer> testSet(FContext ctx, Set<Integer> thing) throws TException {
            return thing;
        }

        @Override
        public List<Integer> testList(FContext ctx, List<Integer> thing) throws TException {
            return thing;
        }

        @Override
        public Numberz testEnum(FContext ctx, Numberz thing) throws TException {
            return thing;
        }

        @Override
        public long testTypedef(FContext ctx, long thing) throws TException {
            return thing;
        }

        @Override
        public Map<Integer, Map<Integer, Integer>> testMapMap(FContext ctx, int hello) throws TException {
            Map<Integer, Integer> mp1 = new HashMap<>();
            mp1.put(-4,-4);
            mp1.put(-3,-3);
            mp1.put(-2,-2);
            mp1.put(-1,-1);

            Map<Integer, Integer> mp2 = new HashMap<>();
            mp2.put(4,4);
            mp2.put(3,3);
            mp2.put(2,2);
            mp2.put(1,1);

            Map<Integer, Map<Integer, Integer>> rMapMap = new HashMap<>();
            rMapMap.put(-4, mp1);
            rMapMap.put(4, mp2);
            return rMapMap;
        }

        @Override
        public boolean TestUppercaseMethod(FContext ctx, boolean thing) throws TException {
            return thing;
        }

        @Override
        public Map<Long, Map<Numberz, Insanity>> testInsanity(FContext ctx, Insanity argument) throws TException {
            Map<Numberz, Insanity> mp1 = new HashMap<>();
            mp1.put(Numberz.findByValue(2), argument);
            mp1.put(Numberz.findByValue(3), argument);

            Map<Numberz, Insanity> mp2 = new HashMap<>();

            Map<Long, Map<Numberz, Insanity>> returnInsanity = new HashMap<>();
            returnInsanity.put((long) 1, mp1);
            returnInsanity.put((long) 2, mp2);

            return returnInsanity;
        }

        @Override
        public Xtruct testMulti(FContext ctx, byte arg0, int arg1, long arg2, Map<Short, String> arg3, Numberz arg4, long arg5) throws TException {
            Xtruct r = new Xtruct();

            r.string_thing = "Hello2";
            r.byte_thing = arg0;
            r.i32_thing = arg1;
            r.i64_thing = arg2;

            return r;
        }

        @Override
        public void testException(FContext ctx, String arg) throws TException {
            switch (arg) {
                case "Xception":
                    Xception e = new Xception();
                    e.errorCode = 1001;
                    e.message = arg;
                    throw e;
                case "TException":
                    throw new TException("Just TException");
                default:
            }
        }

        @Override
        public Xtruct testMultiException(FContext ctx, String arg0, String arg1) throws TException {
            switch (arg0) {
                case "Xception":
                    Xception e = new Xception();
                    e.errorCode = 1001;
                    e.message = "This is an Xception";
                    throw e;
                case "Xception2":
                    Xception2 e2 = new Xception2();
                    e2.errorCode = 2002;
                    e2.struct_thing = new Xtruct();
                    e2.struct_thing.string_thing = "This is an Xception2";
                    throw e2;
                default:
                    Xtruct r = new Xtruct();
                    r.string_thing = arg1;
                    return r;
            }
        }

        @Override
        public void testOneway(FContext ctx, int secondsToSleep) throws TException {
        }
}
